Add 2.18 type inference workaround from change log to sound-problems.md (#4196)
dart-lang/site-www: src/_guides/language/sound-problems.md

> ## Troubleshooting

Reviewer: It looks like it was this way before, but our style guidelines want at least one sentence between headers.

Dart enforces a sound type system. Roughly, this means
you can't get into a situation where the value stored in a variable is
different from the variable's static type. Like most modern statically
typed languages, Dart accomplishes this with a combination of
[static (compile-time)](#static-errors-and-warnings) and [dynamic (runtime)](#runtime-errors) checking.

For example, the following type error is detected at compile-time:

{:.fails-sa}
{% prettify dart tag=pre+code %}
List<int> numbers = [1, 2, 3];
List<String> [!string = numbers!];
{% endprettify %}

Since neither `List<int>` nor `List<String>` is a subtype of the other,
Dart rules this out statically. 

> You can see other examples of static analysis errors, as well as other error types, in the sections below.

Reviewer: ```suggestion
You can see other examples of static analysis errors,
as well as other error types, in the following sections.
```

It's a holdover from the old days, but I do try to avoid "above" and "below" as, if this is translated to Chinese or whatever, that isn't relevant.

## Troubleshooting

> Dart enforces a sound type system. Roughly, this means
> you can't get into a situation where the value stored in a variable is
> different from the variable's static type. Like most modern statically
> typed languages, Dart accomplishes this with a combination of
> [static (compile-time)](#static-errors-and-warnings) and [dynamic (runtime)](#runtime-errors) checking.

Reviewer 2: **suggestion:** This could be trimmed down.
```suggestion
Dart enforces a sound type system. In general terms, this means you can't write code where a variable's value differs from its static type. An `int` type can't store a number with a decimal place. Dart checks variable values against their types at [compile-time](#static-errors-and-warnings) and [runtime](#runtime-errors).
```

These changes drop readability from Grade 11 to Grade 5. Sure, it's eight more words, but they're small.

Reviewer 3: As this PR didn't write this text and there are some particularities of these adjustments that may need further review, perhaps best to leave these changes for a separate PR. 

The whole article could likely use some readability improvements, so perhaps that could be done in a follow-up PR?

## Troubleshooting

> Dart enforces a sound type system. This means you can't write code where a
> variable's value differs from its static type.
> An `int` type can't store a number with a decimal place.
> Dart checks variable values against their types at
> [compile-time](#static-errors-and-warnings) and [runtime](#runtime-errors).

Reviewer 3: Clarifies it's not the `int` type that stores a value, but rather a variable.

```suggestion
Dart enforces a sound type system. 
This means you can't write code where a
variable's value differs from its static type.
A variable with an `int` type can't store a number with a decimal place.
Dart checks variable values against their types at
[compile-time](#static-errors-and-warnings) and [runtime](#runtime-errors).
```

<hr>

### Incorrect type inference

```nocode
error - ...

> ```

Author: @mit-mit @kevmoo I would like to add the actual error message that would show up with the example mentioned in the [change log](https://github.com/dart-lang/sdk/blob/master/CHANGELOG.md#language-1) to match the rest of the page. I could not find it though, so I'm wondering the best way to find relevant prs when starting from the change log. 

I followed the link in the change log to the issue, but couldn't locate the actual PR where the error message was added to the code (if that error message was actually added as part of 2.18, as it being mentioned in the change log implies).

Reviewer 3: My recommendation with these is to include the code in a project and then run `dart analyze` to see the resulting error. It's a bit hard to follow the code implemented in the SDK, as errors touch a lot of pieces.

In this case, with the code you included below, I get an error like the following:

```
error - The return type '...' isn't a '...', as required by the closure's context - return_of_invalid_type_from_closure
```

Eventually, the goal is to get the code in to an example(https://github.com/dart-lang/site-www/tree/main/examples/type_system) so we can keep track of the actual error in an analysis results file(https://github.com/dart-lang/site-www/blob/main/examples/type_system/analyzer-results-stable.txt) and use the code excerpt tool when referencing it. That way if the error message updates we'll know to update this. This is a bit of a pain to set up currently though for error messages, especially ones not yet in stable, so it's fine if we add that later.

> On rare occasions, Dart's enhaced type inference may infer the wrong type for function literal arguments in a generic constructor invocation.

Reviewer: ```suggestion
On rare occasions, Dart's enhanced type inference might infer
the wrong type for function literal arguments in a generic constructor invocation.
```

Line length!!!  :D. We don't break at exactly 80 chars, we try to break on phrasing.

Reviewer 3: https://github.com/dart-lang/site-shared/blob/main/doc/writing-for-dart-and-flutter-websites.md#semantic-line-breaks

This primarily affects `Iterable.fold`

#### Example

> In this code, type inference will infer that `a` has a type of `Null`:

Reviewer: ```suggestion
In the following code,
type inference infers that `a` has a type of `Null`:
```

{:.fails-sa}
<?code-excerpt "lib/... ()" replace=""?>
{% prettify dart tag=pre+code %}

> void main() {

Reviewer: Yeah, once you modify code in an excerpt, you have to run "./tool/refresh-excerpts" in your branch. But, if you have any questions about code excerpts, Khanh is our expert!

Author: I definitely have a lot of questions about the excerpts haha. I'll reach out!

> void main() {

Reviewer 3: Perhaps we don't need to include the `main()` function scaffolding in these example, so it's easier to focus on the relevant code.

<hr>

### Incorrect type inference

> ```nocode
> error - The return type '...' isn't a '...', as required by the closure's context - return_of_invalid_type_from_closure
> ```
> 

Reviewer 3: Okay I ran the analyzer again with Dart 2.18 and 2.19 and I'm getting different errors. Perhaps I was using an older development version of Dart on that machine.

Since this has to do with type inference, it results in many different errors depending on the context, due to types flowing through the code in different ways according to each situation.

Perhaps it doesn't make a lot of sense to include a specific error due to this, and your explanation of "incorrect type inference" should be sufficient. What do you think?

```suggestion
```

Author: Definitely agree, especially now when we just need to get the update out for 2.18. Thanks so much!

On rare occasions, Dart's enhanced type inference might infer
the wrong type for function literal arguments in a generic constructor invocation.
This primarily affects `Iterable.fold`.

#### Example

In the following code,
type inference will infer that `a` has a type of `Null`:

{:.fails-sa}

> <?code-excerpt "lib/common_fixes_analysis.dart (type-inf-null)"?>

Reviewer 3: You'll need to rerun the code excerpt tool again after making changes to the excerpt files.
